lib/engines/content_block_manager/app/services/content_block_manager/find_and_replace_embed_codes_service.rb

```diff
@@ -0,0 +1,37 @@
+module ContentBlockManager
+  class FindAndReplaceEmbedCodesService
+    def self.call(html)
+      new(html).call
+    end
+
+    def call
+      embed_content_references.each do |reference|
+        content_block = content_blocks.find { |c| c.document.content_id == reference.content_id }
+        next if content_block.nil?
+
+        html.gsub!(reference.embed_code, content_block.render)
+      end
+
+      html
+    end
+
+  private
+
+    attr_reader :html
+
+    def initialize(html)
+      @html = html
+    end
+
+    def embed_content_references
+      @embed_content_references ||= ContentBlockTools::ContentBlockReference.find_all_in_document(html)
+    end
+
+    def content_blocks
+      @content_blocks ||= ContentBlockManager::ContentBlock::Edition.joins(:document).where(
```

Comment:
This query would return all published Editions (of which there could be many per Document because we don't have a `superseded` workflow) but I think we would only want the `latest_edition` scope?
I'm wondering if the `render` should belong to the Document, as the embed code isn't going to change throughout its lifecycle, so we will always only be referring to the `live` content right?

Reply:
otherwise when you update an edition the changes wouldn't propagate to other content blocks, or am I misunderstanding?

Reply:
Good point on the latest edition scope. I did consider having it on the document, but I thought we might want to render draft blocks too (for example, when doing the in-app preview)

Reply:
> we might want to render draft blocks too (for example, when doing the in-app preview)

ah yeah that's true... I guess as long as this service and the API only ever returns the live one then it will be ok, would be worth adding a Document with a couple of Editions in the test

Reply:
I've tweaked it so we have a `current_versions` scope now, which ensures we're only querying the latest published editions.